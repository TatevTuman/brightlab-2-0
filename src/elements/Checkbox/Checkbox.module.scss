@import "static/assets/styles/variables";

$size: toRem(20);

.checkbox {
  display: block;
  height: $size;
  position: relative;
  padding-left: 30px;
  margin-bottom: 5px;
  cursor: pointer;
  font-size: 1rem;
  user-select: none;

  &:hover, &:focus {
    outline: none;

    input {
      & ~ .checkbox-indicator {
        background: var(--secondary-dark);
      }

      &:checked ~ .checkbox-indicator {
        background: var(--grey);
      }
    }
  }

  &-indicator {
    position: absolute;
    top: 2px;
    left: 0;
    height: $size;
    width: $size;
    background: var(--grey);
    border: 0 solid #000000;
    border-radius: 5px;
    transition: all 0.2s ease;

    &:after {
      display: none;
      box-sizing: unset;
      content: '';
      position: absolute;
      left: toRem(8);
      top: toRem(4);
      width: toRem(3);
      height: toRem(8);
      border: solid #ffffff;
      border-width: 0 toRem(2) toRem(2) 0;
      transform: rotate(45deg);
    }
  }

  input {
    position: absolute;
    z-index: -1;
    opacity: 0;

    &:checked ~ .checkbox-indicator {
      background: var(--secondary-dark);

      &:after {
        display: block;
      }
    }

    &:disabled ~ .checkbox-indicator {
      background: var(--grey);
      opacity: 0.6;
      pointer-events: none;
    }
  }

  //&-validation-error {
  //  margin-left: -30px;
  //}
}
