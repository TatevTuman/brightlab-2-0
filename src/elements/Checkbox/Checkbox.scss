@import "static/assets/styles/interactive";
@import "static/assets/styles/mixins";

.checkbox {
  @include disabled($background: false, $opacity: true, $pointerEvents: true);

  display: flex;
  position: relative;
  cursor: pointer;
  user-select: none;

  input {
    background-color: var(--white);
    border: 2px solid var(--purple);
    transition: border 0.2s ease, background-color 0.2s ease;
    will-change: border, background-color;

    &:hover {
      background-color: var(--white-pst);
      border: 2px solid var(--purple-d);
    }

    &:active,
    &:checked {
      background-color: var(--purple-dst);
      border: 2px solid var(--purple-dst);
    }

    &[data-error='true'] {
      @include background_border(white, danger);
    }

    appearance: none;
    border-radius: 4px;
    width: 18px;
    height: 18px;

    background-size: 70%;
    background-position: center;
    background-repeat: no-repeat;

    &:checked {
      background-image: url("/assets/images/checkbox.svg");
    }

    margin-right: toRem(12);
    cursor: pointer;
  }

  label {
    font-size: toRem(15);
    font-weight: 400;
    color: var(--primary);

    a {
      font-weight: 600;
    }
  }
}
