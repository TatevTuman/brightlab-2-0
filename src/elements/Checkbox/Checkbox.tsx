import React, { memo } from 'react'
import { ValidationProps } from '@types'
import styles from './Checkbox.module.scss'

export interface CheckboxProps extends ValidationProps {
  name: string
  label?: string
  checked?: boolean
  disabled?: boolean
  onChange?: (value: boolean) => void
}

const Checkbox: React.FC<CheckboxProps> = props => {
  const { name, label, checked, disabled, register, validation, errors, onChange } = props

  const isRequired = !!validation?.required
  const ref = register && register(validation)
  const error = errors && errors[name]

  return (
    <label className={styles.checkbox}>
      {label && (
        <label htmlFor={name} data-required={isRequired}>
          {label}
        </label>
      )}
      <input
        id={name}
        type="checkbox"
        name={name}
        ref={ref}
        checked={checked}
        disabled={disabled}
        onChange={e => onChange && onChange(!e.currentTarget.checked)}
      />
      <div className={styles.checkboxIndicator} />
      {error && <span className={`${styles.checkboxValidationError} validation-error`}>{error.message}</span>}
    </label>
  )
}

export default memo(Checkbox)
