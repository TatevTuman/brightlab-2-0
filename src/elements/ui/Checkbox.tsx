import React, { memo, forwardRef } from 'react'
import cls from 'classnames'
import { ClassName } from '~types'
import * as Icon from '~svg'

interface CheckboxProps {
  className?: ClassName
  name: string
  label?: string
  value: boolean
  defaultValue?: boolean
  onChange: (value: boolean) => void
}

const Checkbox: React.FC<CheckboxProps> = forwardRef<HTMLInputElement, CheckboxProps>((props, ref) => {
  const { className, name, label, value, defaultValue, onChange } = props

  return (
    <div className={'inline-flex'}>
      {label && (
        <label className={'mr-8'} htmlFor={name}>
          {label}
        </label>
      )}
      <input
        className={'hidden'}
        id={name}
        name={name}
        ref={ref}
        type="checkbox"
        checked={value}
        defaultChecked={defaultValue}
        onChange={() => onChange(!value)}
      />
      <label
        htmlFor={name}
        className={cls(
          'w-16 h-16 border border-green-1000 rounded transition-colors duration-150 hover:bg-green-100',
          { 'bg-green-1000 bg-checkbox hover:bg-green-1000': value },
          className
        )}
      >
        {value && <Icon.Check className={'text-white'} />}
      </label>
    </div>
  )
})

Checkbox.displayName = 'Checkbox'

export default memo(Checkbox)
