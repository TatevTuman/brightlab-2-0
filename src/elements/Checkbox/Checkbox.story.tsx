import React from 'react'
import { Meta, Story } from '@storybook/react/types-6-0'
import Checkbox, { CheckboxProps } from './Checkbox'
import { storybook } from '@utils'

interface CheckboxStoryTemplateProps extends CheckboxProps {}

export default {
  title: 'Elements/Checkbox',
  argTypes: {
    label: { control: 'text', name: 'Label' },
    checked: { control: 'boolean', name: 'Checked' },
    ...storybook.controls.interactive,
    name: storybook.args.disabled
  }
} as Meta

const CheckboxStoryTemplate: Story<CheckboxStoryTemplateProps> = args => {
  return <Checkbox {...args} />
}

CheckboxStoryTemplate.args = {
  name: 'Checkbox',
  label: 'Text',
  ...storybook.actions.interactive
}

export const CheckboxStory = CheckboxStoryTemplate.bind({})
CheckboxStory.args = {
  ...CheckboxStoryTemplate.args
}
