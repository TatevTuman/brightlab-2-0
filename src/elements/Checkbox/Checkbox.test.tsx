import React from 'react'
import { cleanup, fireEvent, render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Checkbox from './Checkbox'

// beforeAll(() => {})
// afterAll(() => {})
// beforeEach(() => {})
afterEach(() => {
  cleanup()
  jest.clearAllMocks()
})

describe('Checkbox', () => {
  it('renders correctly', async () => {
    const { container } = render(<Checkbox name={'checkbox'} checked={true} />)
    expect(container).toMatchSnapshot()
  })

  it('renders no label correctly', () => {
    const { container } = render(<Checkbox name={'checkbox'} checked={true} />)

    const label = container.querySelector('label')?.querySelector('label')

    expect(label).toBe(undefined)
  })

  it('renders label correctly', () => {
    const props = {
      name: 'checkbox',
      label: 'checkbox label',
      checked: true
    }

    const { getByText } = render(<Checkbox {...props} />)

    expect(getByText(props.label)).toBeInTheDocument()
  })

  it('renders required label correctly', () => {
    const props = {
      name: 'checkbox',
      label: 'checkbox label',
      required: true,
      checked: true
    }

    const { getByText } = render(<Checkbox {...props} />)
    const label = getByText(props.label)

    expect(label).toBeInTheDocument()
    expect(label).toHaveAttribute('data-required', 'true')
  })

  it('handles onChange on label', () => {
    const props = {
      name: 'checkbox',
      label: 'checkbox label',
      checked: true,
      onChange: jest.fn()
    }

    const { getByText, getByLabelText } = render(<Checkbox {...props} />)
    const label = getByText(props.label)
    const checkbox = getByLabelText(props.label) as HTMLInputElement

    userEvent.click(label!)

    expect(props.onChange).toHaveBeenCalledTimes(1)
    expect(checkbox.checked).toBeTruthy()
  })

  it('handles onChange', () => {
    const props = {
      name: 'checkbox',
      label: 'checkbox label',
      checked: true,
      onChange: jest.fn()
    }

    const { getByLabelText } = render(<Checkbox {...props} />)
    const checkbox = getByLabelText(props.label) as HTMLInputElement

    userEvent.click(checkbox!)

    expect(props.onChange).toHaveBeenCalledTimes(1)
    expect(checkbox.checked).toBeTruthy()
  })

  it('handles onKeyDown', async () => {
    const props = {
      name: 'checkbox',
      label: 'checkbox label',
      checked: true,
      onKeyDown: jest.fn()
    }

    const { getByLabelText } = render(<Checkbox {...props} />)
    const checkbox = getByLabelText(props.label) as HTMLInputElement

    fireEvent.keyDown(checkbox, { key: 'A' })
    expect(props.onKeyDown).toHaveBeenCalledTimes(1)
  })
})
