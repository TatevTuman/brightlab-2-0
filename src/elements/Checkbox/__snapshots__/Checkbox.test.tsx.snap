// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders correctly 1`] = `
<div>
  <div
    class="checkbox undefined"
  >
    <input
      checked=""
      data-testid="checkbox"
      id="checkbox"
      name="checkbox"
      tabindex="0"
      type="checkbox"
    />
  </div>
</div>
`;
